### ISSUE TITLE
`needless_pass_by_ref_mut` false positive

### ISSUE THREAD

**Issue by iamsauravsharma** [rust-lang/rust-clippy#11179]
### Summary

I got
```
error: this argument is a mutable reference, but not used mutably
  --> src/migrator/any.rs:22:17
   |
22 |     connection: &mut <Any as sqlx::Database>::Connection,
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider changing to: `&<Any as sqlx::Database>::Connection`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_pass_by_ref_mut
note: the lint level is defined here
  --> src/lib.rs:3:9
   |
3  | #![deny(clippy::all)]
   |         ^^^^^^^^^^^
   = note: `#[deny(clippy::needless_pass_by_ref_mut)]` implied by `#[deny(clippy::all)]`
```
even when connection is used as mutable https://github.com/iamsauravsharma/sqlx_migrator/blob/master/src/migrator/any.rs#L36
 GitHub actions error https://github.com/iamsauravsharma/sqlx_migrator/actions/runs/5583309753/jobs/10203493083

### Lint Name

needless_pass_by_ref_mut

### Reproducer

1. clone repo at https://github.com/iamsauravsharma/sqlx_migrator
2. run cargo clippy --features any,postgres you will see clippy error
```
error: this argument is a mutable reference, but not used mutably
  --> src/migrator/any.rs:22:17
   |
22 |     connection: &mut <Any as sqlx::Database>::Connection,
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider changing to: `&<Any as sqlx::Database>::Connection`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_pass_by_ref_mut
```
Expects no error since connection is used mutably and using as suggestion doesn't compile code

### Version

```text
rustc 1.73.0-nightly (da6b55cc5 2023-07-17)
binary: rustc
commit-hash: da6b55cc5eaf76ed6acb7dc2f7d611e32af7c9a7
commit-date: 2023-07-17
host: aarch64-apple-darwin
release: 1.73.0-nightly
LLVM version: 16.0.5
```

**Comment by Centri3** [comment]
cc @GuillaumeGomez

**Comment by y21** [comment]
Minimal reproducible example:
```rs
async fn f(x: &mut i32) {
  *x += 1;
}
```
Looks like it doesn't work correctly in async fns.

**Comment by iamsauravsharma** [comment]
Also, It has no issue if you add extra parameter(except String)
```
// no issue
async fn _f(x: &mut i32, _p: i32) {
    *x += 1;
}
```
If all other parameter is string it have issue
```
async fn _g(x: &mut i32, _p: String) {
    *x += 1;
}

async fn _h(x: &mut i32, _p: String, _q: String) {
    *x += 1;
}
```

**Comment by iamsauravsharma** [comment]
Also, It works properly if you add extra parameter which are not stored in heap? (not sure)
```
// no issue
async fn _f(x: &mut i32, _p: i32) {
    *x += 1;
}
and doesn't work with heap based parameter? (not sure)
```
If all other parameter is string it doesn't work
```
async fn _g(x: &mut i32, _p: String) {
    *x += 1;
}

async fn _h(x: &mut i32, _p: Vec<i32>, _q: Box<u64>) {
    *x += 1;
}
```

**Comment by GuillaumeGomez** [comment]
Thanks! Taking a look.

**Comment by GuillaumeGomez** [comment]
I opened https://github.com/rust-lang/rust-clippy/pull/11184 to fix this issue.
